src/glossary.md

```diff
@@ -1,5 +1,29 @@
 # Glossary
 
+## Terms used in the Specification
+
+### Ill-formed Program
+
+A Rust Program is ill-formed if it violates a statically set constraint placed on it by the language.
+Unless otherwise statement explicitly an implementation shall issue an appropriate diagnostic for an ill-formed program and shall not translate such a program.
+
+#### No Diagnostic Required
```

Comment:
Sorry, I'm still struggling with understanding what this means. 

Is this trying to say that some rules may result in a *warning*, but otherwise isn't enforced?

If so, I don't think we have yet agreed on an approach for how to document warnings. In general, we have avoided *all* warnings in the reference. There are a few exceptions when it is directly relevant to the language (like the diagnostic chapter), or when we have felt the warning is very important (like `unsafe_op_in_unsafe_fn`, though that is arguable if it is appropriate.

Reply:
Lints are separate from this.
The compiler can issue a hard error (fail to compile), warn (lint, fcw, etc.), or silently accept it. This isn't for things that you'd lint on, though, this is particularily for things make horribly break the program, but can't be easily (or at all) detected and diagnosed. For example, `asm!(".section .data");`, which can cause the rest of the function (and possibly even entire other functions) to be emitted in the `.data` section, far far away from where the code is expected to be.

Reply:
Sorry, I'm still not understanding what this means. I don't see why there would be rules that we want to say "this may or may not compile". 

I'm trying to understand, is this related to undefined behavior?

I see in the assembly chapter, `asm.invocation.prefix-instr` mentions this. I'm confused because the original text had that in a list of things the programmer must not do to avoid undefined behavior. The new text doesn't seem to be organized that way, and I'm having a hard time understanding why that changed or how it relates to the original text.

Reply:
It's basically super UB. With UB it's fine to have it in the program as long as it's not evaluated*. With IF-NDR, it's not fine to have it in the program.

That particular list (in the original ASM chapter) is particularly bad, because it had quite a few things mixed together, including things that obviously are not fine in unevaluated code because they have potentially invalid global implications - for example, `asm!(".section .data")` as mentioned, which switches to a section called `.data` and depending on how inline asm is implemented, that could affect just the rest of the block, all inline assembly in the crate, or potentially even the rest of the function and all others in the crate. For the same reason, trying to put a function into `.data` is likely fundamentally broken, because depending on the platform, the binary format, and the linker, that could end up in an RWX section, an RW section, and RX section (and mutable data is placed in the same RX section), or outright being rejected by the linker or loader (whether or not the function is even called).